### ISSUE TITLE
Any standard way to include the modified resource for relationship operations?

### ISSUE THREAD

**Issue by cmeeren** [json-api/json-api#1492]
A great benefit of JSON:API is that the updated resource can be included in the response of `PATCH` requests, thus updating the client's representation (e.g. if an `updatedAt` field is changed).

Unfortunately, I can't find a way to do this for to-many relationship operations (`POST`/`PATCH`/`DELETE` to add/replace/remove members). At best, the related resources can be included, but if these don't have a relationship back to the resource that was modified, it seems that a separate `GET` request must be performed to get the updated resource.

For example, assuming that an article can have several authors, `POST /articles/1/relationships/authors` adds an author to the article, but since a `user` resource (the author) does not have a relationship back to the modified `article`, there is no way to include it (and it may have been modified in other ways, e.g. the article's `updatedAt` field may have changed). The only solution for the client to get the updated `article` seems to be to perform a separate `GET /articles/1` request after the relationship update was successful.

Is this correct, or am I missing something? Is this something that could be addressed in a future JSON:API version?

**Comment by cmeeren** [comment]
Perhaps a leading dot could be used? E.g. `?include=.` to include the article in the example above, or `?include=.comments` to include the article and its comments. Seems like an elegant and backward compatible way of supporting this. It fits with the remarks in https://github.com/json-api/json-api/issues/1491#issuecomment-662276631.

**Comment by jelhan** [comment]
Current v1.1 of the specification seems to forbid including the updated relationship(s) in the response if they haven't changed:

> 200 OK
>
> If a server accepts an update but also changes the targeted relationship(s) in other ways than those specified by the request, it MUST return a 200 OK response. The response document MUST include a representation of the updated relationship(s).
>
> A server MUST return a 200 OK status code if an update is successful, the client’s current data remain up to date, and the server responds only with top-level meta data. In this case the server MUST NOT include a representation of the updated relationship(s).

I'm especially referring to the second paragraph. It could be rephrased as: The server must not include a representation of the updated relationship(s) if the client's current data remain up to date. It's not entirely clear what is referred to as "client's current data" but in the context it seems to talk only about the targeted relationship(s).

Not saying that this is good. I just wanted to point out that this part of the specification is confusing. Especially as the specification has a note that `include` query parameter can be used when creating a relationship:

> Note: This section applies to any endpoint that responds with primary data, regardless of the request type. For instance, a server could support the inclusion of related resources along with a POST request to create a resource or relationship.

**Comment by cmeeren** [comment]
I agree that it is confusing. I also think the spec is too strict if it mandates that data should not be returned if unchanged. There are several reasons for this:

* It may be difficult or impossible for general JSON:API framework implementations to know whether or not the data has changed, and thus whether or not data should be returned
* Not responding with up-to-date is an optimization and is not something the spec should force (premature optimization)
* Always responding with data makes it easier for many kinds of clients; it makes the communication more stateless
* Responding with up-to-date data will never break clients

For these reasons, if the spec actually mandates that no data be returned in some cases, I will likely willingly break that part of the spec for simplicity for my framework [Felicity](https://github.com/cmeeren/Felicity).

In any case, this is all tangential to the issue at hand here.

**Comment by jelhan** [comment]
I fully agree that there are good reasons to return data even if the targeted relationship(s) remain unchanged. It also seems to contradict the intend of the specification.

It was added in #657. The intend of that merge request was to allow meta only responses. I don't think it was the intend to forbid returning primary data. It seems as if this potential interpretation of the added sentence was missed in the review.

I just read the sentence again and noticed that it could be also interpreted as tautological:

> A server MUST return a 200 OK status code if [...] the server responds only with top-level meta data. In this case the server MUST NOT include a representation of the updated relationship(s).

If a server response only with top-level meta data, the response document does not include any primary data. Otherwise the server response would not only contain top-level meta data. But if the server response does not include primary data it can not include a representation of the updated relationships(s).

I _think_ this could be fixed within the specification without breaking backward compatibility.

**Comment by cmeeren** [comment]
> I _think_ this could be fixed within the specification without breaking backward compatibility.

I'm happy you think so, and I hope it will be clarified in the spec. However, let me again reiterate that this is tangential to the original issue, which is about the ability to return the modified resource for POST/PATCH/DELETE to the resource's relationship links.

**Comment by cmeeren** [comment]
Is this something that could be considered for a future version of the spec (perhaps even 1.1), or are there fundamental problems with it?

Please see the OP; the rest of the discussion above is tangential.

**Comment by jelhan** [comment]
_I originally replied wrongly at #1567 to this proposal. Please excuse the confusion._

> Is this something that could be considered for a future version of the spec (perhaps even 1.1), or are there fundamental problems with it?

The proposal to use `.` as a keyword to identify target resource in a relationship update seems reasonable to me. See https://github.com/json-api/json-api/issues/1492#issuecomment-662275093 for more details on that proposal.

I tried to write a pull request for it myself in https://github.com/json-api/json-api/compare/gh-pages...jelhan:patch-15. Doing so I noticed that it would violate the full linkage requirement for compound documents.

> Compound documents require “full linkage”, meaning that every included resource MUST be identified by at least one [resource identifier object](https://jsonapi.org/format/#document-resource-identifier-objects) in the same document. These resource identifier objects could either be primary data or represent resource linkage contained within primary or included resources.

For a relationship operation the response contains resource identifier objects of the related resources only. It does not contain the resource identifier object of the target resource itself. At least that is how I understand the specification regarding the confusion wording (see https://github.com/json-api/json-api/issues/1524#issuecomment-1166288191 for details on that one).

To simplify let's have a look at `GET` requests for a relationship link for now. The specification includes the following example for `GET /articles/1/relationships/author`:

> ```http
> HTTP/1.1 200 OK
> Content-Type: application/vnd.api+json
> 
> {
>   "links": {
>     "self": "/articles/1/relationships/author",
>     "related": "/articles/1/author"
>   },
>   "data": {
>     "type": "people",
>     "id": "12"
>   }
> }
> ```

Inclusion of the related author could be requested using `?include=author`. The response payload would change to something like this:

> ```http
> HTTP/1.1 200 OK
> Content-Type: application/vnd.api+json
> 
> {
>   "links": {
>     "self": "/articles/1/relationships/author?include=author",
>     "related": "/articles/1/author"
>   },
>   "data": {
>     "type": "people",
>     "id": "12"
>   },
>   "included": [
>     {
>       "type": "people",
>       "id": "12",
>       "attributes": {
>         "name": "John Doe"
>       }
>     }
>   ]
> }
> ```

Everything fine so far.

Let's have a look on inclusion of the article itself. Following the proposal the client would request doing so using `?include=.`. The server would response with a payload like this:

> ```http
> HTTP/1.1 200 OK
> Content-Type: application/vnd.api+json
> 
> {
>   "links": {
>     "self": "/articles/1/relationships/author?include=.",
>     "related": "/articles/1/author"
>   },
>   "data": {
>     "type": "people",
>     "id": "12"
>   },
>   "included": [
>     {
>       "type": "articles",
>       "id": "1",
>       "attributes": {
>         "title": "Getting specifications right is challenging"
>       }
>     }
>   ]
> }
> ```

The included resource `{ "type": "articles", "id": "1" }` would _not_ be identified by any resource identifier object included as primary data. I fear that's a fundamental problem.

**Comment by cmeeren** [comment]
> I fear that's a fundamental problem.

I don't agree.

It breaks the current spec, sure. But suppose an exception to full linkage was added for this use-case. There is already an exception to full linkage wrt. sparse fieldsets. So assume we change this:

> The only exception to the full linkage requirement is when relationship fields that would otherwise contain linkage data are excluded via sparse fieldsets.

to this:

> The only exceptions to the full linkage requirement are:
> * when relationship fields that would otherwise contain linkage data are excluded via sparse fieldsets
> * when including the owning resource of a relationship using `include=.` in operations against a relationship's `self` URL

(Of course a more carefully considered phrasing would be in order; the above is just to get my point across.)

I don't see that change as a fundamental problem. It would then allow this very useful feature of always being able to return the resource that owns a relationship when you update the relationship, regardless of whether you replace the relationship data using `PATCH` to the resource endpoint (in which case you get it as the primary data) or `PATCH`/`POST`/`DELETE` to the relationship `self` endpoint (which currently has no guaranteed way to do this; it requires the relationship resources to have a relationship back to the parent and even then would pollute the response with all the relationship's resources).

**Comment by jelhan** [comment]
I agree that it may be solvable introducing a new exception for full linkage requirement. But I think we should be careful about it as it may increase complexity a lot.

A client would not be able to identify the target resource in `included` only looking at the response. To do so, the client would need to understand the URL naming schema used by the API. Or the client would need to analyze how it came up with the URL used in the relationship operation and identify the resource based on that.

This seems to break some fundamental design patterns around JSON:API specification. A client would need additional information to process a response.

I'm not saying that this must be a showstopper. But we should carefully consider if we want to except these trade-offs.

I would recommend to gather some experience using an [extension](https://jsonapi.org/format/1.1/#extensions) before adding to the base specification. Such an extension could be designed similar to the base specification, which may even help to reduce implementation complexity.

The extension could define two additional members of the JSON:API document:

- `self:data`
- `self:included`

`self:data` is a JSON:API resource object representing the target resource in a relationship operation. E.g. for `GET /articles/1/relationships/author` it would be the resource `{ "type": "articles", "id": "1" }`.

`self:included` is a list of related resources of the resource included as `self:data`. Full linkage requirement would apply to `self:data` and `self:included` as well. Every resource in `self:included` must be either directly or indirectly linked to `self:data`.

Furthermore, the extension could register a query parameter `self:include` allowing a client to request inclusion of related resources for `self:data`. For example a client could request inclusion of all comments of an article when updating the article's author using a request like this: `GET /articles/1/relationships/author?self:include=comments`

To avoid including the same resource twice, once in `included` and once in `self:included`, the extension may want to enforce an empty value for `include` query parameter.

**Comment by dgeb** [comment]
I haven't had the time to personally understand and process this request (or @jelhan's response) yet but have added it to the [v1.1 milestone](https://github.com/json-api/json-api/milestone/2) to ensure that it's not overlooked before finalizing.

**Comment by dgeb** [comment]
Thanks for your patience as I've been mulling this over.

To make sure we're all on the same page, this is the context from [section 8.3: Inclusion of Related Resources](http://localhost:9876/format/1.1/#fetching-includes):

> Furthermore, related resources can be requested from a relationship endpoint:
> 
> ```http
> GET /articles/1/relationships/comments?include=comments.author HTTP/1.1
> Accept: application/vnd.api+json
> ```
> 
> In this case, the primary data would be a collection of
> resource identifier objects that represent linkage to comments for an article,
> while the full comments and comment authors would be returned as included data.

@cmeeren My understanding of this request is that, in the context of the above example, you'd like to allow a request to include `comments` and the related `article`, in this case the resource with id `1`. This can be achieved with by referencing `comments.article` as an include:

```
GET /articles/1/relationships/comments?include=comments.author,comments.article
```

And you clearly understand this based on your statement and example:

> At best, the related resources can be included, but if these don't have a relationship back to the resource that was modified, it seems that a separate GET request must be performed to get the updated resource.
>
> For example, assuming that an article can have several authors, POST /articles/1/relationships/authors adds an author to the article, but since a user resource (the author) does not have a relationship back to the modified article, there is no way to include it (and it may have been modified in other ways, e.g. the article's updatedAt field may have changed). The only solution for the client to get the updated article seems to be to perform a separate GET /articles/1 request after the relationship update was successful.

So in the context of your example, you could make the request:

```
POST /articles/1/relationships/authors?include=authors.articles
```

But if the `author` related resource is actually a `user` which doesn't have a relationship to `articles`, then the full linkage requirement fails and the author's articles can not also be returned.

I imagine we're all in agreement so far. 

However, what concerns me is that the following request would exhibit the same issue:

```
GET /authors/1?include=articles
```

Again, we're talking about the same resources and the same lack of linkage, yet we are not making a request to a relationship endpoint. I'm wary of introducing an exception to the full linkage requirement that would apply for one of these requests but not the other. Moreover, both of these scenarios could be solved through resource design, such that there is an `author` (or perhaps `user:author`) resource that either extends or references a base `user`, and also has an `articles` relationship.

I will consider this further and will be glad to discuss further, although I don't think it's a blocker for v1.1. And of course, as @jelhan suggested, please feel free to explore an extension that tackles this problem and let us know how it goes if you do so!

**Comment by cmeeren** [comment]
> Moreover, both of these scenarios could be solved through resource design, such that there is an `author` (or perhaps `user:author`) resource that either extends or references a base `user`, and also has an `articles` relationship.

I have two significant issues with this:

* It seems like a _very_ heavy-handed solution just to get the current resource being modified. I can't imagine I would ever do this for that reason alone.
* The `articles` relationship would presumably contain _all_ the user's (possibly hundreds, or thousands) of articles, whereas my point here is just to be able to include _the updated representation of the resource that was actually modified_. That resource can't easily  be modeled by a relationship from a user(-derived) resource, simply because it does not exist in the context of a user (or author); it exists only in the context of the request that is being performed.

Let me try to rephrase.

To be entirely clear, my _only_ issue here is that when updating a relationship via the relationship's `self` link, many kinds of clients would find it very useful to be able to get the updated representation of the resource.

This should be possible without having to include it through a potentially large to-many relationship. In `GET /articles/1/relationships/comments?include=comments.article`, if there are hundreds of comments, they are just unnecessary dead weight in the response if the client is just interested in, say, the article's updated `numComments` attribute.

I am not talking about making exceptions to full linkage for anything other than the resource update operations. I'm not sure how the `GET /authors/1?include=articles` operation you are concerned about is relevant (in light of my clarifications above).

My specific suggestion is that in

```
POST /articles/1/relationships/authors
```

if I want to include the modified article, I am proposing to use `include=.`. A leading dot in the include path is a "self reference" and would include the current resource. This only makes sense for relationship endpoints. (In all other cases, this would be the primary data anyway and be included in the response. It can in those cases be blocked with an error response, or perhaps effectively ignored.)

So

```
POST /articles/1/relationships/authors?include=.
```

would return the relationship's identifier objects as the primary data (as usual), and the modified resource (the `article` with ID `1`) as included data (with no linkage pointing to it).

What kind of API clients would this be useful for? Certainly the common type that is front-ends with a client-side in-memory store of resources, keyed by `type` and `id`. Whenever a resource is received, the received representation is used to update the currently stored representation, and the UI updates with the new information. Additionally, it would be useful for API clients that know about the URL structure or what resource it is currently modifying. Together, I think that covers most real-world clients. (JSON:API isn't full HATEOAS / REST level 3 anyway.) While we should in general be careful about making assumptions about the nature of API clients, I don't think we should let over-generalizations stand in the way of features that are useful for many clients and irrelevant (but not detrimental) to some.

---

Side-note: In my experience, the lack of this feature is detrimental even to unrelated requests. When clients must request the updated resource in a separate request when updating to-many relationships, I have often seen them just request the resource separately even for normal `PATCH` resource requests, where the response includes the updated resource automatically and the extra request therefore is completely unnecessary. This is presumably done to reduce the number of code paths to get an updated resource representation, but of course unnecessarily doubles the number of requests in write-heavy APIs.

**Comment by jelhan** [comment]
> To be entirely clear, my _only_ issue here is that when updating a relationship via the relationship's `self` link, many kinds of clients would find it very useful to be able to get the updated representation of the resource.

I'm wondering if this (and similar use cases for other side effects due to creating, updating, or deleting resources and relationships) could be better solved with an extension.

An `effects` extension could define an `effects:data` extension top-level member. Server would use it to inform the client about side-effects.

`effects:data` would take an array of effects objects. Each effect object describes one side-effect. It has two properties:

1. `effect`, which value is either `created`, `modified` or `deleted`.
2. `data`, which is a resource object if `effect` is `created` or `modified` and a resource identifier object if `effect` is `deleted`.

Such an extension might solve the describes use case better:

1. No strict full linkage requirement. It would even support side-effects on resources, which do not have a relationship with target resource.
2. No over-fetching as server would only include modified resources. And may even only include sparse fieldset of modified fields of a resource.
3. It would support _any_ resource or relationship modification.

Using compound documents and `include` query parameter to inform clients of side-effects always felt like a hack to me. v1.1 seems to enable better alternatives using extensions.

**Comment by cmeeren** [comment]
That's not a bad idea. I will consider that. (Realistically, probably not any time soon, though.)
